### PR TITLE
[Refactoring] [Testing] [Feature] NoiseHandler

### PR DESCRIPTION
**Pull request by chMoussa**
Supersedes #584  and solves #583 (besides plotting). We introduce a `NoiseHandler` class containing information about the noise. For digital noise, this means now that we can do a conversion in `convert_ops` to the right class in Pyq. @jpmoutinho also proposed to separate noise protocol types into Analog, Digital, and Readout (given that a same protocol name may be available in different settings such as depolarizing).

Note that a substantial harmonisation of naming may be needed between different backend (Readout is basically called SPAM in Pulser).

Todos:

- [x] Redesign the previous `Noise` class into the `NoiseHandler` class to allow multiple combination of noise protocols.
- [x] Change typings and replace `DigitalNoise` in blocks.
- [x] Add conversion function for Pyq backend.
- [x] Serialize noise in blocks
- [x] Add Tests
- [x] Elaborate the docs.

### CODE REVIEW

File: qadence/mitigations/readout.py [hunk 1]
```diff
@@ -88,7 +88,12 @@ def mitigation_minimization(
     Returns:
         Mitigated counts computed by the algorithm
     """
-    noise_matrices = noise.options.get("noise_matrix", noise.options["confusion_matrices"])
+    readoutnoise = noise if isinstance(noise, NoiseSource) else noise.noise_sources[-1]
```
Comment (RolandMacDoland) [inline qadence/mitigations/readout.py:91]:
```suggestion
    readout_noise = noise if isinstance(noise, NoiseSource) else noise.noise_sources[-1]
```

File: qadence/mitigations/readout.py [hunk 2]
```diff
@@ -88,7 +88,12 @@ def mitigation_minimization(
...
+    if readoutnoise.protocol != "Readout":
```
Comment (RolandMacDoland) [inline qadence/mitigations/readout.py:92]: Can't you compare to a `NoiseType` enum instead of a hardcoded `str` ?

File: qadence/noise/protocols.py [hunk 3]
```diff
@@ -1,29 +1,62 @@
...
-    DEPHASING = "dephasing"
-    DEPOLARIZING = "depolarizing"
-    READOUT = "readout"
+class NoiseSource:
+    """A container for a single source of noise."""
+
+    def __init__(self, protocol: str, options: dict = dict(), protocol_type: str = "") -> None:
+        if protocol not in supported_noise_protocols:
+            raise ValueError(
+                "Protocol {protocol} is not supported. Choose from {supported_noise_protocols}."
```
Comment (RolandMacDoland) [inline qadence/noise/protocols.py:23]:
```suggestion
                f"Protocol {protocol} is not supported. Choose from {supported_noise_protocols}."
```

File: qadence/noise/protocols.py [hunk 4]
```diff
@@ -33,27 +66,161 @@ def get_noise_fn(self) -> Callable:
...
+        return {
+            "protocol": self.protocol,
+            "options": self.options,
+            "protocol_type": self.protocol_type,
+        }
 
     @classmethod
-    def _from_dict(cls, d: dict) -> Noise | None:
+    def _from_dict(cls, d: dict | None) -> NoiseSource | None:
         if d:
```
Comment (RolandMacDoland) [inline qadence/noise/protocols.py:77]:
```suggestion
        if d is not None:
```

File: qadence/noise/protocols.py [hunk 5]
```diff
@@ -33,27 +66,161 @@ def get_noise_fn(self) -> Callable:
...
+    def __repr__(self) -> str:
+        return f"{self.protocol_type} NoiseSource({self.protocol}, {str(self.options)})"
+
+
+class NoiseHandler:
+    """A container for multiple sources of noise."""
+
+    def __init__(
+        self,
+        protocol: str | NoiseSource | list[str] | list[NoiseSource],
```
Comment (RolandMacDoland) [inline qadence/noise/protocols.py:95]: Do we still need `str` ?

File: qadence/noise/protocols.py [hunk 6]
```diff
@@ -33,27 +66,161 @@ def get_noise_fn(self) -> Callable:
...
+        else:
+            protocol = protocol if isinstance(protocol, list) else [protocol]
+            options = options if isinstance(options, list) else [options] * len(protocol)
+            types = (
+                protocol_type
+                if isinstance(protocol_type, list)
+                else [protocol_type] * len(protocol)
+            )
+
+            if len(options) != len(protocol) or len(types) != len(protocol):
```
Comment (RolandMacDoland) [inline qadence/noise/protocols.py:113]: There might be noises that do not require options ? In that case, you'll fall in the first of you predicate cases without necessarily being wrong ?

File: qadence/mitigations/readout.py (hunk 2 above)
Comment (RolandMacDoland) [reply qadence/mitigations/readout.py:92]: For instance `NoiseProtocolType` ?

File: qadence/types.py [hunk 7]
```diff
@@ -459,3 +459,28 @@ class ExperimentTrackingTool(StrEnum):
 
 
 LoggablePlotFunction = Callable[[Module, int], tuple[str, Figure]]
+
+
+class NoiseProtocolType(StrEnum):
```
Comment (RolandMacDoland) [inline qadence/types.py:464]: Would it make sense to merge this two enums somehow ? So that you could do smtg like `NoiseType.ANALOG.DEPOLARIZING` ?

File: tests/qadence/test_noise/test_digital_noise.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    hamiltonian_factory,
+    kron,
+    set_noise,
+)
+
+list_noises = [DigitalNoiseType(noise.value) for noise in DigitalNoiseType]
+
+
+@pytest.mark.parametrize("protocol", list_noises)
+@given(st.digital_circuits())
```
Comment (RolandMacDoland) [inline tests/qadence/test_noise/test_digital_noise.py:26]: Nice :)

File: tests/qadence/test_noise/test_digital_noise.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    NoiseHandler,
+    QuantumCircuit,
+    QuantumModel,
+    Z,
+    hamiltonian_factory,
+    kron,
+    set_noise,
+)
+
+list_noises = [DigitalNoiseType(noise.value) for noise in DigitalNoiseType]
```
Comment (RolandMacDoland) [inline tests/qadence/test_noise/test_digital_noise.py:22]: Since it is reused in more than one test, maybe it could become a fixture ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks a lot for this @chMoussa ! My own impression is that there is a profusion of `StrEnum` that maybe we should think about simplifiying, `eg.` `NoiseType`, `NoiseProtocoltype` etc. It'll rapidly get out of hand if we don't try to reduce the number of them. Otherwise it looks good. I would also try out some deterministic test with noise just to make sure it gives correct values back. I'll also seek for @jpmoutinho and @mlahariya feedback.

File: qadence/types.py (hunk 7 above)
Comment (chMoussa) [reply qadence/types.py]: Made a dataclass `NoiseProtocol` with different levels of noises.

File: qadence/noise/protocols.py (hunk 6 above)
Comment (chMoussa) [reply qadence/noise/protocols.py]: I am forcing noise without options to be an empty dict

File: qadence/noise/protocols.py (hunk 5 above)
Comment (chMoussa) [reply qadence/noise/protocols.py]: Not anymore with the new definition of NoiseProtocol

File: tests/qadence/test_noise/test_digital_noise.py (hunk 9 above)
Comment (chMoussa) [reply tests/qadence/test_noise/test_digital_noise.py]: So far, better to keep the parameterize on the list instead of doing a for loop in the test.

File: docs/tutorials/realistic_sims/noise.md [hunk 10]
```diff
@@ -1,7 +1,37 @@
...
+One can also define a `NoiseHandler` as a list of `NoiseSource` instances:
+
+```python exec="on" source="material-block" session="noise" result="json"
+from qadence import NoiseHandler, NoiseSource
+from qadence.types import NoiseProtocol
+
+digital_noise = NoiseSource(protocol=NoiseProtocol.DIGITAL.DEPOLARIZING, options={"error_probability": 0.1})
+readout_noise = NoiseSource(protocol=NoiseProtocol.READOUT.READOUT, options={"error_probability": 0.1, "seed": 0})
+
+noise_combination = NoiseHandler([digital_noise, readout_noise])
```
Comment (RolandMacDoland) [inline docs/tutorials/realistic_sims/noise.md:28]: If the `NoiseHandler` can accept a single protocol (as shown in line 61 below) why would it need a list of `NoiseSource` ? It feels a bit on an unnecessary intermediate step unless there is a reason I don't see now.

File: docs/tutorials/realistic_sims/noise.md [hunk 11]
```diff
@@ -69,16 +98,45 @@ print(f"noiseless = {noiseless_exp}") # markdown-exec: hide
...
 
 ```python exec="on" source="material-block" session="noise" result="json"
-from qadence import NoiseType, DigitalNoise, RX, run
+from qadence import NoiseProtocol, RX, run
 import torch
 
-noise = DigitalNoise(NoiseType.BITFLIP, error_probability = 0.2)
-noise = DigitalNoise.bitflip(error_probability = 0.2) # equivalent
+noise = NoiseHandler(NoiseProtocol.DIGITAL.BITFLIP, {"error_probability": 0.2})
+noise = NoiseHandler.bitflip({"error_probability": 0.2}) # equivalent
```
Comment (RolandMacDoland) [inline docs/tutorials/realistic_sims/noise.md:139]: Would it make sense to build on this to chain noise models ? Smtg like:

```python
noise = NoiseHandler.bitflip(...).phaseflip(...).<other-noise-model>(...)
```

That keep adding in the list.

File: pyproject.toml [hunk 12]
```diff
@@ -26,7 +26,7 @@ authors = [
 ]
 requires-python = ">=3.9"
 license = { text = "Apache 2.0" }
-version = "1.7.8"
+version = "1.7.9"
```
Comment (RolandMacDoland) [inline pyproject.toml:29]: I think you can bump the minor.

File: qadence/backends/pulser/backend.py [hunk 13]
```diff
@@ -235,27 +235,27 @@ def run(
...
-        if noise_probs is None:
-            KeyError("A `noise probs` option should be passed to the <class QuantumModel>.")
-        if not (isinstance(noise_probs, float) or isinstance(noise_probs, Iterable)):
-            KeyError(
-                "A single or a range of noise probabilities"
-                " should be passed. Got {type(noise_probs)}."
-            )
+        noise_source = noise.noise_sources[-1]
+        if not isinstance(noise_source.protocol, NoiseProtocol.ANALOG):
+            raise TypeError("Noise must be Analog.")
```
Comment (RolandMacDoland) [inline qadence/backends/pulser/backend.py:246]:
```suggestion
            raise TypeError("Noise must be of type `NoiseProtocol.ANALOG`.")
```

File: qadence/backends/pyqtorch/convert_ops.py [hunk 14]
```diff
@@ -286,17 +300,35 @@ def convert_block(
...
     else:
         raise NotImplementedError(
             f"Non supported operation of type {type(block)}. "
             "In case you are trying to run an `AnalogBlock`, make sure you "
             "specify the `device_specs` in your `Register` first."
         )
+
+
+def convert_digital_noise(noise: NoiseHandler | None) -> pyq.noise.NoiseProtocol | None:
+    if noise is None:
```
Comment (RolandMacDoland) [inline qadence/backends/pyqtorch/convert_ops.py:324]: Maybe do this check outside this function ? So that you don't have to do it every time it is called and you won't have to handle the `None` return type ?

File: qadence/noise/protocols.py [hunk 15]
```diff
@@ -1,29 +1,37 @@
...
 
 PROTOCOL_TO_MODULE = {
-    "readout": "qadence.noise.readout",
+    "Readout": "qadence.noise.readout",
 }
 
-# Temporary solution
-DigitalNoise = NoiseProtocol
 
+class NoiseSource:
```
Comment (RolandMacDoland) [inline qadence/noise/protocols.py:13]: I have the feeling all the checking `NoiseSource` provides can be done in the handler. Is that true ?

File: qadence/types.py [hunk 16]
```diff
@@ -459,3 +461,29 @@ class ExperimentTrackingTool(StrEnum):
...
+
+
+class AnalogNoise(StrEnum):
+    """Type of noise protocol."""
+
+    DEPOLARIZING = "Depolarizing"
+    DEPHASING = "Dephasing"
+
+
+class ReadoutNoise(StrEnum):
```
Comment (RolandMacDoland) [inline qadence/types.py:473]: Yeah unsure we need to maintain an `StrEnum` for a single variant.

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa more comments from me as I _think_ we can simplify further. Another point is about testing for composed noise models. Have you tried it out ?

File: docs/tutorials/realistic_sims/noise.md (hunk 10 above)
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md:28]: This is especially for the case of Digital noise where we can combine several types (if the user decide to apply let's say a bitflip followed by a phase flip, etc)

File: docs/tutorials/realistic_sims/noise.md (hunk 11 above)
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md:139]: Originally I intented for these methods to instantiate easily a handler with one type of noise. Maybe I could add a `append` method to add to the list?

File: qadence/noise/protocols.py (hunk 15 above)
Comment (chMoussa) [reply qadence/noise/protocols.py:13]: I wanted to separate the checking of options at the NoiseSource level, and for the handler to check more the combination of several types.

File: qadence/types.py (hunk 16 above)
Comment (chMoussa) [reply qadence/types.py:473]: At the moment no, but there may be other readout types implemented at a later stage.

File: docs/tutorials/realistic_sims/mitigation.md [hunk 17]
```diff
@@ -45,7 +45,7 @@ observable = hamiltonian_factory(circuit.n_qubits, detuning=Z)
 model = QuantumModel(circuit=circuit, observable=observable)
 
 # Define a noise model to use:
-noise = Noise(protocol=Noise.READOUT)
+noise = NoiseHandler(protocol=NoiseProtocol.READOUT.READOUT)
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply docs/tutorials/realistic_sims/mitigation.md:48]: Made it work by setting `NoiseProtocol.READOUT` to be just a string for now.

File: qadence/types.py (hunk 16 above)
Comment (chMoussa) [reply qadence/types.py]: But I will replace it for now and do this change when we introduce another readout protocol

**Comment by chMoussa** [comment]
> Thanks @chMoussa more comments from me as I _think_ we can simplify further. Another point is about testing for composed noise models. Have you tried it out ?

Yes, I just added a parameterized test for this in `qadence/tests/qadence/test_noise/test_digital_noise.py`

File: docs/tutorials/realistic_sims/noise.md [hunk 18]
```diff
@@ -1,7 +1,37 @@
...
+noise_combination = NoiseHandler([digital_noise, readout_noise])
+print(noise_combination)
+```
+
+!!! warning "NoiseHandler scope"
+    Note it is not possible to define a `NoiseHandler` instances with both digital and analog noise sources, both readout and analog noise sources, several analog noise sources, several readout noise sources, or a readout noise source that is not the last defined `NoiseSource` within `NoiseHandler`.
+
 ## Readout errors
 
 State Preparation and Measurement (SPAM) in the hardware is a major source of noise in the execution of
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md:37]: NoiseHandler.bitflip would only create an instance with only one noisesource.

File: qadence/mitigations/analog_zne.py [hunk 19]
```diff
@@ -116,11 +116,12 @@ def noise_level_experiment(
     circuit: QuantumCircuit,
     observable: list[AbstractBlock],
     param_values: dict[str, Tensor],
-    noise: Noise,
+    noise: NoiseHandler,
     endianness: Endianness,
     state: Tensor | None = None,
 ) -> Tensor:
-    noise_probs = noise.options.get("noise_probs")
+    noise_source = noise.noise_sources[-1]
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/mitigations/analog_zne.py:123]: You cannot create a NoiseHandler with empty list. You get a raised error when instantating a NoiseHandler. And I made the convention that, if a readout is specified, it has to be the last noisesource (also raises an error at initialization).

File: docs/tutorials/realistic_sims/noise.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md:52]: And just added

File: docs/tutorials/realistic_sims/noise.md (hunk 11 above)
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md:164]: Added the `append`

**Review by jpmoutinho — APPROVED** [review]
Very nice @chMoussa, this looks much cleaner :)

I did not go into the full details of the code, let me know if there is something specific that you would like to discuss. But in general, this interface has my approval.

File: qadence/backends/pulser/backend.py [hunk 20]
```diff
@@ -24,12 +24,12 @@
 from qadence.measurements import Measurements
 from qadence.mitigations import Mitigations
 from qadence.mitigations.protocols import apply_mitigation
-from qadence.noise import Noise
-from qadence.noise.protocols import apply_noise
+from qadence.noise import NoiseHandler
+from qadence.noise.protocols import apply_readout_noise
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/backends/pulser/backend.py:28]: Yes

**Comment by chMoussa** [comment]
> Thanks @chMoussa. Sorry for being a pain but I think we still have too many types that makes the whole implementation difficult to follow. Ideally, I _think_ we can get away with max two types. Also, would it make sense to turn the `NoiseHandler` into a pydantic model for all the checking ? It might also help in serialization. Just a suggestion.

@RolandMacDoland, I followed your suggestions. Now we only have the `NoiseHandler` and the classmethods are becoming mutation methods. Let me know what you think of it

File: qadence/backends/pulser/backend.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/backends/pulser/backend.py:28]: And in a future MR following a Pyqtorch release, we will have an interface ReadoutNoise that should instantiate the corresponding backend object.

File: qadence/noise/protocols.py [hunk 21]
```diff
@@ -33,27 +52,231 @@ def get_noise_fn(self) -> Callable:
...
+    def _from_dict(cls, d: dict | None) -> NoiseHandler | None:
+        if d is not None and d.get("protocol", None):
+            return cls(d["protocol"], d["options"])
         return None
 
     @classmethod
     def list(cls) -> list:
         return list(filter(lambda el: not el.startswith("__"), dir(cls)))
 
+    def filter(self, protocol_type: NoiseEnum) -> NoiseHandler | None:
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/noise/protocols.py]: Changed to `protocol`

File: qadence/noise/protocols.py (hunk 15 above)
Comment (chMoussa) [reply qadence/noise/protocols.py]: I get it, now it's removed.

File: docs/tutorials/realistic_sims/noise.md (hunk 10 above)
Comment (chMoussa) [reply docs/tutorials/realistic_sims/noise.md]: Gotcha, now it's been reworked.
